Allow electron process forks of modules that use pre-gyp

bcomnes opened the pull request:
closes https://github.com/mapbox/node-pre-gyp/issues/278

I still need to check to see if this works in built electron apps, but this is fairly close to solving the issue of allowing `childProcess.fork`s in electron of node-pre-gyp modules.

springmeyer commented:
Thanks for this PR @bcomnes and especially for the detailed description of the problem over at #278. I'm the maintainer of node-pre-gyp and happy to accept this PR once there are tests. I'm not an electron user so I'd love help on how to best set up tests to confirm this is working and ensure it never regresses (which will run on travisci). Are writing unit tests something you could do? Or could you advise on how to test manually and maybe others could help convert those into unit tests?

bcomnes commented:
👌 I'll put something together today.

bcomnes commented:
A decent test would be to basically try to load a node-pre-gyp module in a module running in a childProcess.fork of electron.  Would you be opposed to pulling in https://github.com/electron-userland/electron-prebuilt as a dev dependency?

bcomnes commented:
Yarg, why is that failing

bcomnes commented:
The other weird thing about this is requiring electron.  Ideally we could pull off the correct electron version from the thing that actually called the fork, rather than an action-at-a-distancey require resolve of a package.json.

bcomnes commented:
Oh crap, I just pushed some unintentional files up.  Fixing.

bcomnes commented:
Ok I made a few changes:

- Simplified the electron case of get_process_runtime
- No longer mutate process.versions.  
- The electron version check happens in get_runtime_abi instead
- I opened an issue upstream to electron to expose the version of electron to the forked electron process somehow https://github.com/electron/electron/issues/9058
- Added a mock test to test out the detection of ELECTRON_RUN_AS_NODE

Is that enough testing for this you think? Or do we want to add more of an acceptance / integration test closer to what https://github.com/ballpit/electron-sqlite3-fork-bug/blob/master/main.js tests for?

bcomnes commented:
@springmeyer any feedback on the tests/current state of the PR?

bcomnes commented:
Rebased.

bcomnes commented:
@springmeyer Do you have any feedback on the tests I added?

springmeyer commented on test/electron/index.js:
```diff
@@ -0,0 +1,13 @@
+        assert.equal(1,1);
+        var opts = evaluate(require('../app1/package.json'));
+        console.log(opts)
+        console.log(find(path.resolve(path.join(__dirname,'../app1/package.json'))))
```
This should assert something rather than just `console.log`, correct?

bcomnes replied:
These snuck in as well.  Fixing.

bcomnes replied:
This was a leftover from an attempt to actually perform an integration test with electron, but there are some costs to adding those in like adding additional CI prebuilt downloads.  I decided to just go with the unit tests in `test/versioning.test.js` which seem to cover the issue at play here.  Removing this file for now.

springmeyer commented on lib/util/versioning.js.orig:
```diff
@@ -0,0 +1,317 @@
+"use strict";
```
I presume you added this `lib/util/versioning.js.orig` by accident? Can you remove and rebase?

bcomnes replied:
Derp.  Not sure how this snuck in here.  Removing the orig file.

springmeyer commented:
@bcomnes - made a request for a few minor changes. If you can address those then I am willing to merge. I admit that I don't fully understand the changes or how they work. So I would love if anyone else that uses electron could weight in.

bcomnes commented:
Ok, cleaned up those two issue, and rebased.  

To TLDR explain the issue: when forking a process in electron, a special ENV var is introduced to the forked process environment (`ELECTRON_RUN_AS_NODE`) and any prebuilt native dependencies need to pick this up in order to find the correct bindings, otherwise the normal checks that node-pre-gyp performs to check if its running electron fail and it returns the vanilla node bindings (or at least this is one approach).  

The only real weirdness here is the electron version resolution.  Ideally this would also come in from the ENV of the forked process, but its not available yet.  I'm currently working with the electron team to get that in (see https://github.com/electron/electron/issues/9058) and will PR that improved method once its available.

bcomnes commented:
Hrmmm tests are failing now.  Any idea if its because of what I'm doing?

springmeyer commented:
> Any idea if its because of what I'm doing?

@bcomnes - I think they may be unrelated. Can you rebase against the master branch and see if things are fixed?

bcomnes commented:
Rebased.  Fingers crossed!

bcomnes commented:
In a spat of inspiration to make the world better this morning, I squashed and rebased this.

An alternative PR for the same problem is here: https://github.com/mapbox/node-pre-gyp/pull/343

It looks like the difference between the two is ^^ that one requires you to pass in an ENV var yourself, this PR attempts to resolve the correct electron version for you.   Personally, I think this is how it *should* work, but the code is a little more ugly until we can support from upstream electron.

Getting rock solid sqlite support in electron would be amazing, and this would help make that happen.

bcomnes commented:
All tests passing locally.  CI seems to be failing on unrelated tests 🤔 any ideas?

bcomnes commented:
@springmeyer any clue whats going on in the failed runs?  Also what happened to travis?

dimalo commented:
Hey guys, thanks for your work! Please make this merge finally happen!

springmeyer commented:
Thanks so much @bcomnes. Would love to land this. At the same time I don't have time or ability to review or merge currently. Nor debug travis, but it might be something like https://github.com/mapbox/gzip-hpp/issues/33 which resulted from .org -> .com issues I think.

codecardinal commented:
@springmeyer I need this PR for sqlite3 too :P

JakeRadMSFT commented:
@bcomnes it looks like electron made the change you wanted: 
https://github.com/electron/electron/issues/8727

bcomnes commented:
@JakeRadMSFT The thing that I wanted to remove was https://github.com/mapbox/node-pre-gyp/pull/279/files#diff-314ae8d53157a25e4458ced25e711105R87

https://github.com/electron/electron/pull/16450 adds a `process.versions.electron` which suposedly provides that.  Let me try it out.

bcomnes commented:
Ok, re-reading this.  Totally untested, but: IF `process.versions.electron` is added into forks, as noted in electron/electron#9058, then this PR is no longer needed.  Since tests arn't passing on master, I dont even think the `process.env.ELECTRON_RUN_AS_NODE` part is needed.

Closing.

awohletz commented:
Confirmed that updating to Electron v6 (currently in beta) fixed the issue for me. E.g. I am able to use node-sqlite3 from an Electron-forked process.
